src/main/java/com/minecolonies/coremod/entity/ai/citizen/enchanter/EntityAIWorkEnchanter.java

```diff
@@ -0,0 +1,377 @@
+            }
+            else
+            {
+                worker.swingArm(EnumHand.OFF_HAND);
+            }
+            return getState();
+        }
+
+        final int slot = InventoryUtils.findFirstSlotInItemHandlerWith(new InvWrapper(worker.getInventoryCitizen()), IS_ANCIENT_TOME);
+        if (slot != -1)
+        {
+            final ICitizenData data = worker.getCitizenData();
+            if (data != null)
+            {
+                final int openSlot = InventoryUtils.getFirstOpenSlotFromItemHandler(new InvWrapper(worker.getInventoryCitizen()));
+                if (openSlot == -1)
+                {
+                    //Dump if there is no open slot.
+                    incrementActionsDone();
+                    progressTicks = 0;
+                    return IDLE;
+                }
+
+                final Tuple<ItemStack, Integer> tuple = IColonyManager.getInstance().getCompatibilityManager().getRandomEnchantmentBook(getOwnBuilding().getBuildingLevel());
+
+                data.spendLevels(tuple.getSecond());
+                worker.getCitizenExperienceHandler().updateLevel();
+                worker.getInventoryCitizen().setInventorySlotContents(openSlot,tuple.getFirst());
+
+                InventoryUtils.reduceStackInItemHandler(new InvWrapper(worker.getInventoryCitizen()), new ItemStack(ModItems.ancientTome));
+                incrementActionsDoneAndDecSaturation();
+            }
+        }
+        progressTicks = 0;
+        return IDLE;
+    }
+
+    /**
+     * Gather experience from a worker.
+     * Go to the hut of the worker.
+     * Wait for the worker.
+     * Drain, and then return to work building.
+     * @return next state to go to.
+     */
+    private IAIState gatherAndDrain()
+    {
+        if (job.getPosToDrainFrom() == null)
+        {
+            return IDLE;
+        }
+
+        if (walkToBlock(job.getPosToDrainFrom()))
+        {
+            return getState();
+        }
+
+        final AbstractBuildingWorker buildingWorker = getOwnBuilding().getColony().getBuildingManager().getBuilding(job.getPosToDrainFrom(), AbstractBuildingWorker.class);
+        if (buildingWorker == null)
+        {
+            resetDraining();
+            getOwnBuilding(BuildingEnchanter.class).removeWorker(job.getPosToDrainFrom());
+            return IDLE;
+        }
+
+        if (citizenToGatherFrom == null)
+        {
+            final List<Optional<AbstractEntityCitizen>> workers = buildingWorker.getAssignedEntities().stream()
+                                                                    .filter(e -> e.isPresent() && e.get().getCitizenData().getLevel() > 0)
+                                                                    .collect(Collectors.toList());
+            final Optional<AbstractEntityCitizen> citizen;
+            if (workers.size() > 1)
+            {
+                citizen = workers.get(worker.getRandom().nextInt(workers.size()));
```

Comment:
Thought: possibly add feature for higher level worker or etc to specifically seek out workers with high experience to make the process more efficient.

Reply:
No?

Reply:
Not for now, to be fair, he is fairly efficient already, don't want him too OP =D